### ISSUE TITLE
Syncing 0 Bytes files

### ISSUE THREAD

**Issue by krakazyabra** [nextcloud/server#23107]
### Steps to reproduce
1. Make sync between NC and folder on PC
2. Upload file with 0 bytes size through web interface
3. Try to sync

### Expected behaviour
The file will be synced from NC to PC

### Actual behaviour
Error while syncing

### Server configuration

**Operating system:**
Debian 10 Buster
**Web server:**
nginx/1.19.1
**Database:**
Percona xtradb-cluster 5.7.19
**PHP version:**
PHP 7.3.20
**Nextcloud version:** (see Nextcloud admin page)
19.0.1-fpm
**Updated from an older Nextcloud/ownCloud or fresh install:**
Rolling updates from 16 
**Where did you install Nextcloud from:**
 Docker image
**Signing status:**
<details>
<summary>Signing status</summary>

```
No messages
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.5.0
  - activity: 2.12.0
  - admin_audit: 1.9.0
  - appointments: 1.7.6
  - apporder: 0.11.0
  - bookmarks: 3.4.0
  - calendar: 2.0.4
  - camerarawpreviews: 0.7.8
  - checksum: 0.4.5
  - cloud_federation_api: 1.2.0
  - comments: 1.9.0
  - contacts: 3.3.0
  - contactsinteraction: 1.0.0
  - cookbook: 0.7.6
  - dashboardcharts: 0.1.3
  - dav: 1.15.0
  - deck: 1.0.5
  - dicomviewer: 1.2.2
  - encryption: 2.7.0
  - external: 3.6.0
  - extract: 1.2.4
  - federatedfilesharing: 1.9.0
  - federation: 1.9.0
  - files: 1.14.0
  - files_accesscontrol: 1.9.1
  - files_automatedtagging: 1.9.0
  - files_downloadactivity: 1.8.0
  - files_external: 1.10.0
  - files_markdown: 2.3.1
  - files_mindmap: 0.0.22
  - files_pdfviewer: 1.8.0
  - files_rightclick: 0.16.0
  - files_sharing: 1.11.0
  - files_texteditor: 2.11.0
  - files_trashbin: 1.9.0
  - files_versions: 1.12.0
  - files_videoplayer: 1.8.0
  - firstrunwizard: 2.8.0
  - forms: 2.0.4
  - gpxedit: 0.0.13
  - gpxmotion: 0.0.11
  - gpxpod: 4.2.2
  - groupfolders: 7.0.0
  - groupquota: 0.1.4
  - keeweb: 0.6.3
  - logreader: 2.4.0
  - lookup_server_connector: 1.7.0
  - mail: 1.4.1
  - maps: 0.1.6
  - news: 14.2.2
  - notes: 3.6.4
  - oauth2: 1.7.0
  - ocr: 6.0.56
  - onlyoffice: 6.0.0
  - password_policy: 1.9.1
  - phonetrack: 0.6.4
  - photos: 1.1.0
  - privacy: 1.3.0
  - provisioning_api: 1.9.0
  - quicknotes: 0.6.0
  - quota_warning: 1.8.0
  - recommendations: 0.7.0
  - serverinfo: 1.9.0
  - settings: 1.1.0
  - sharebymail: 1.9.0
  - sociallogin: 3.3.0
  - socialsharing_email: 2.1.0
  - spreed: 9.0.4
  - support: 1.2.1
  - survey_client: 1.7.0
  - systemtags: 1.9.0
  - text: 3.0.1
  - twofactor_backupcodes: 1.8.0
  - updatenotification: 1.9.0
  - viewer: 1.3.0
  - workflowengine: 2.1.0
Disabled:
  - nextcloud_announcements
  - notifications
  - theming
  - theming_customcss
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "theme": "mytheme",
        "blacklisted_files": [],
        "mail_smtpmode": "sendmail",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": 25,
        "mail_smtptimeout": 10,
        "mail_smtpsecure": "",
        "mail_smtpauth": false,
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "logfile": "\/dev\/stdout",
        "enable_previews": true,
        "preview_max_filesize_image": 4096,
        "preview_max_y": 4096,
        "preview_max_x": 4096,
        "enabledPreviewProviders": [
            "OC\\Preview\\PNG",
            "OC\\Preview\\JPEG",
            "OC\\Preview\\GIF",
            "OC\\Preview\\HEIC",
            "OC\\Preview\\BMP",
            "OC\\Preview\\XBitmap",
            "OC\\Preview\\MP3",
            "OC\\Preview\\TXT",
            "OC\\Preview\\MarkDown",
            "OC\\Preview\\Illustrator",
            "OC\\Preview\\Photoshop"
        ],
        "logtimezone": "Europe\/Prague",
        "social_login_auto_redirect": true,
        "updatechecker": false,
        "has_internet_connection": true,
        "objectstore_multibucket": {
            "class": "\\OC\\Files\\ObjectStore\\S3",
            "arguments": {
                "num_buckets": 10,
                "bucket": "nextcloud-",
                "autocreate": true,
                "key": "KEY",
                "secret": "SECRET",
                "use_ssl": true,
                "hostname": "HOSTNAME",
                "port": 443,
                "use_path_style": true
            }
        },
        "onlyoffice": {
            "jwt_header": "Authorization",
            "verify_peer_off": true
        },
        "loglevel": 3,
        "trusted_domains": [
            "nc.domain.org",
            "localhost",
            "10.*.*.*",
            "onlyoffice.domain.org"
        ],
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "filelocking.enabled": false,
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": "6379",
            "timeout": 0
        },
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "19.0.1.1",
        "overwrite.cli.url": "https:\/\/nc.domain.org",
        "overwriteprotocol": "https",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "updater.release.channel": "stable",
        "maintenance": false
    }
}
```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...
Minio S3 as primary storage
**Are you using encryption:** yes/no
no
**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
Social Login application

#### Web server error log
<details>
<summary>Web server error log</summary>

```
2020/09/30 13:17:23 [error] 30#30: *67119 FastCGI sent in stderr: "PHP message: {"reqId":"qQ0bJpdAQ864AafZdrkv","level":3,"time":"2020-09-30T13:17:23+02:00","remoteAddr":"10.36.129.181","user":"auth-dd2f1a98-72ff-42a8-b69e-d0c997fcf33e","app":"PHP","method":"PUT","url":"/remote.php/webdav/zerofile/zeroFile","message":"hash_final(): supplied resource is not a valid Hash Context resource at /var/www/html/lib/private/Files/Stream/HashWrapper.php#71","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0","version":"19.0.1.1"}PHP message: {"reqId":"qQ0bJpdAQ864AafZdrkv","level":3,"time":"2020-09-30T13:17:23+02:00","remoteAddr":"10.36.129.181","user":"auth-dd2f1a98-72ff-42a8-b69e-d0c997fcf33e","app":"PHP","method":"PUT","url":"/remote.php/webdav/zerofile/zeroFile","message":"Cannot modify header information - headers already sent at /var/www/html/apps/dav/lib/Connector/Sabre/File.php#680","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0","version":"19.0.1.1"}PHP message: {"reqId":"qQ0bJpdAQ864AafZdrkv","level":3,"time":"2020-09-30T13:17:23+02:00","remoteAddr":"10.36.129.181","user":"auth-dd2f1a98-72ff-42a8-b69e-d0c997fcf33e","app":"PHP","method":"PUT","url":"/remote.php/webdav/zerofile/zeroFile","message":"hash_final(): supplied resource is not a valid Hash Context resource at /var/www/html/lib/private/Files/Stream/HashWrapper.php#71","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0","version":"19.0.1.1"}PHP message: {"reqId":"qQ0bJpdAQ864AafZdrkv","level":3,"time":"2020-09-30T13:17:23+02:00","remoteAddr":"10.36.129.181","user":"auth-dd2f1a98-72ff-42a8-b69e-d0c997fcf33e","app":"PHP","method":"PUT","url":"/remote.php/webdav/zerofile/zeroFile","message":"Cannot modify header information - headers already sent at /var/www/html/apps/dav/lib/Connector/Sabre/File.php#680","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0","version":"19.0.1.1"}
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
NOTICE: PHP message: {"reqId":"qQ0bJpdAQ864AafZdrkv","level":3,"time":"2020-09-30T13:17:23+02:00","remoteAddr":"10.36.129.181","user":"auth-dd2f1a98-72ff-42a8-b69e-d0c997fcf33e","app":"PHP","method":"PUT","url":"/remote.php/webdav/zerofile/zeroFile","message":"hash_final(): supplied resource is not a valid Hash Context resource at /var/www/html/lib/private/Files/Stream/HashWrapper.php#71","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0","version":"19.0.1.1"}
NOTICE: PHP message: {"reqId":"qQ0bJpdAQ864AafZdrkv","level":3,"time":"2020-09-30T13:17:23+02:00","remoteAddr":"10.36.129.181","user":"auth-dd2f1a98-72ff-42a8-b69e-d0c997fcf33e","app":"PHP","method":"PUT","url":"/remote.php/webdav/zerofile/zeroFile","message":"Cannot modify header information - headers already sent at /var/www/html/apps/dav/lib/Connector/Sabre/File.php#680","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:80.0) Gecko/20100101 Firefox/80.0","version":"19.0.1.1"}

```
</details>

**Comment by krakazyabra** [comment]
I tried to reproduce this on try.nextcloud.com, but there avaerything is fine and file is syncing in both directions. So, I presume, this is somehow connected with S3. 
In reproduce steps add 0 step: 
Use S3 as primary storage.

**Comment by acsfer** [comment]
Duplicate of https://github.com/nextcloud/server/issues/22029 https://github.com/nextcloud/server/issues/22164 https://github.com/nextcloud/server/issues/22437

Should be fixed in the next release with #22651.
